source/distributed/solution_transfer.cc

```diff
@@ -206,73 +201,81 @@ namespace parallel
-        dofs_per_cell);
-      for (typename std::vector<const VectorType *>::iterator it =
-             input_vectors.begin();
-           it != input_vectors.end();
-           ++it)
+
+      // create buffer for each individual object
+      std::vector<::dealii::Vector<typename VectorType::value_type>> dofvalues(
+        input_vectors.size());
+
+      auto cit_input = input_vectors.cbegin();
+      auto it_output = dofvalues.begin();
+      for (; cit_input != input_vectors.cend(); ++cit_input, ++it_output)
         {
-          cell->get_interpolated_dof_values(*(*it), dofvalues);
-          std::memcpy(data_store,
-                      &dofvalues(0),
-                      sizeof(typename VectorType::value_type) * dofs_per_cell);
-          data_store += dofs_per_cell;
+          it_output->reinit(dofs_per_cell);
+          cell->get_interpolated_dof_values(*(*cit_input), *it_output);
         }
+
+      Utilities::pack(dofvalues, data, /*allow_compression=*/false);
     }
 
 
+
     template <int dim, typename VectorType, typename DoFHandlerType>
     void
     SolutionTransfer<dim, VectorType, DoFHandlerType>::unpack_callback(
       const typename Triangulation<dim, DoFHandlerType::space_dimension>::
         cell_iterator &cell_,
       const typename Triangulation<dim, DoFHandlerType::space_dimension>::
         CellStatus /*status*/,
-      const void *               data,
-      std::vector<VectorType *> &all_out)
+      const boost::iterator_range<std::vector<char>::const_iterator> data_range,
+      std::vector<VectorType *> &                                    all_out)
     {
       typename DoFHandlerType::cell_iterator cell(*cell_, dof_handler);
 
       const unsigned int dofs_per_cell = cell->get_fe().dofs_per_cell;
-      ::dealii::Vector<typename VectorType::value_type> dofvalues(
-        dofs_per_cell);
-      const typename VectorType::value_type *data_store =
-        reinterpret_cast<const typename VectorType::value_type *>(data);
 
-      for (typename std::vector<VectorType *>::iterator it = all_out.begin();
-           it != all_out.end();
-           ++it)
+      std::vector<::dealii::Vector<typename VectorType::value_type>> dofvalues =
+        Utilities::unpack<
+          std::vector<::dealii::Vector<typename VectorType::value_type>>>(
+          data_range.begin(), data_range.end(), /*allow_compression=*/false);
+
+      // check if sizes match
+      Assert(dofvalues.size() == all_out.size(), ExcInternalError());
+
+      // check if we have enough dofs provided by the FE object
+      // to interpolate the transferred data correctly
+      for (auto it_dofvalues = dofvalues.begin();
+           it_dofvalues != dofvalues.end();
+           ++it_dofvalues)
+        Assert(dofs_per_cell <= it_dofvalues->size(),
+               ExcMessage(
+                 "The transferred data was packed on less dofs than the"
+                 "currently registered FE object assigned to the DoFHandler."));
```

Comment:
I didn't know that this behavior is possible or even wanted. But since the below depicted testcases rely on that feature, I wanted to double-check this condition with an assertion.

Reply:
Are you saying that the *finite element* used on this cell has fewer DoFs ("dofs_per_cell") than the one used for packing?

Reply:
Exactly. The order of the FE object differs throughout the transfer process in some testcases.
@bangerth and I agreed on discussing this separately in issue #6880.